### ISSUE TITLE
How to build static libtensorflow_cc.a and libtensorflow_framework.a for Linux and MacOS

### ISSUE THREAD

**Issue by samhodge** [tensorflow/tensorflow#28388]
<em>Please make sure that this is a build/installation issue. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:build_template</em>

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): MacOS 10.13
- TensorFlow installed from (source or binary): source
- TensorFlow version: v1.13.1
- Python version: 3.6.7
- Installed using virtualenv? pip? conda?: conda
- Bazel version (if compiling from source): 0.19.2
- GCC/Compiler version (if compiling from source): Apple LLVM version 9.1.0 (clang-902.0.39.2)
- CUDA/cuDNN version: 10/7.4.2.24
- GPU model and memory: GTX 1060

**Describe the problem**
I am unable to build a static version of tensorflow as described for a solution to:

https://github.com/tensorflow/tensorflow/issues/22810#issuecomment-433225466

**Provide the exact sequence of commands / steps that you executed before running into the problem**

Trying to use tensorflow/contrib/cmake results in a number of errors

**Any other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

Can you describe the approach to creating a static version of the libraries needed for the following headers:

#include <tensorflow/core/platform/init_main.h>
#include <tensorflow/core/public/session.h>
#include <tensorflow/core/framework/tensor_shape.h>

To link against my software?

Or alternatively solve this issue:

https://github.com/tensorflow/tensorflow/issues/22810

Dated in October of last year.

Static linking seems to be doable.

**Comment by muddham** [comment]
@samhodge This question is better asked on [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is not a bug or feature request. There is also a larger community that reads questions there.If you think we've misinterpreted a bug, please comment again with a clear explanation, as well as all of the information requested in the [issue template](https://github.com/tensorflow/tensorflow/issues/new/choose). Thanks!

**Comment by samhodge** [comment]
The feature I am requesting is to build a static library under OSX the makefile contrib does not work

**Comment by samhodge** [comment]
Maybe this is  a better link

https://stackoverflow.com/questions/56013508/libtensorflow-core-a-missing-symbols-tensorflowportinitmainchar-const-in

**Comment by gunan** [comment]
I do not think we ever explored building static libtensorflow libraries.
@martinwicke to confirm if that's the case
@meteorcloudy to check how we can do this with bazel

Martin, also who would be the new owners of libtensorflow?

**Comment by mihaimaruseac** [comment]
I have attempted this for OSSFuzz integration and got into issues with Bazel

**Comment by samhodge** [comment]
I sorted the issues on the Linux side of things and can get everything statically linked with libtensorflow-core.a nsync.a and libprotobuf.a I was missing some of the -W,l directives which I shamelessly stole from the benchmark example that came as part of contrib/makefile

**Comment by meteorcloudy** [comment]
@hlopko Is it possible to write a cc_static_library in Starlark now?

**Comment by hlopko** [comment]
@oquenchil will know for sure :)

**Comment by samhodge** [comment]
OK I can build the static libraries under OSX from the makefile method also, but I dont have CUDA support yet.

I might submit a new Makefile if I figure it out.

**Comment by oquenchil** [comment]
> @hlopko Is it possible to write a cc_static_library in Starlark now?

With help from the toolchain API but yes it can be done now. There is no cc_common.link() or cc_common.archive() method in the API that will create a static library with dependencies.

**Comment by mihaimaruseac** [comment]
Is there a recipe for doing it now?

**Comment by samhodge** [comment]
OK here is the link command for using the libtensorflow-core.a nsync.a and libprotobuf.a on OSX

```
    TENSORFLOW_LIBS = -Wl,-force_load,$(TENSORFLOW)/tensorflow/contrib/makefile/gen/lib/libtensorflow-core.a -Wl,-force_load,$(TENSORFLOW)/tensorflow/contrib/makefile/downloads/nsync/builds/default.macos.c++11/nsync.a  -Wl,-force_load,$(TENSORFLOW)/tensorflow/contrib/makefile/gen/protobuf-host/lib/libprotobuf.a -lstdc++ -lz -lpthread 
```

**Comment by samhodge** [comment]
OK people I can see that under contrib/makefile 

There are CUDA build possibilities for Tegra 

Is there any history of people makeing the tf_gpu.a for Linux and OSX?

I am not sure if my users will buy a CPU only interence because it can take 5 minutes per frame of video.

GPU can be as little as 0.2fps.

I have tested with libtensorflow_cc.so and libtensorflow_framework.so

Sam
sam

**Comment by oquenchil** [comment]
> Is there a recipe for doing it now?

Not exactly. We have this example:
https://github.com/bazelbuild/rules_cc/blob/master/examples/my_c_compile/my_c_compile.bzl

You could use that with CPP_LINK_STATIC_LIBRARY_ACTION_NAME instead from https://github.com/bazelbuild/bazel/blob/master/tools/build_defs/cc/action_names.bzl

**Comment by hlopko** [comment]
There is also this example which is closer to what is needed: https://github.com/bazelbuild/rules_cc/blob/master/examples/my_c_archive/my_c_archive.bzl.

Now you have 2 options:

1) Wait until we add accessors for objects to LibraryToLink
2) Collect transitive static libraries, unpack them, and repack again in one big static library.

**Comment by samhodge** [comment]
@hlopko 

I think option 2 would be suitable.

Sam

**Comment by samhodge** [comment]
Any progress on this, I tried hacking up the Tegra gpu/cuda libtfcuda.a target to work on Linux and I didn't get very far:

see:
https://github.com/tensorflow/tensorflow/blob/v1.13.1/tensorflow/contrib/makefile/Makefile#L833

Has anybody attempted such a thing before?

Seems possible.

Sam

**Comment by samhodge** [comment]
The other problem I am having one of my models uses contributed Ops which are not part of the current Makefile so that doesn't run using libtensorflow-core.a but works fine with libtensorflow_cc.so and libtensorflow_framework.so

I hope I am not the only person who needs static linking.

Sam

**Comment by jesen8** [comment]
@samhodge hi,Dears,,i have got a problem with build a static library and run error ,could u can help me? thanks.

**Comment by jesen8** [comment]
the detail info is [here ](https://stackoverflow.com/questions/56249350/how-to-build-a-static-library-with-libtensorflow-core-a-and-how-to-use-it)

**Comment by samhodge** [comment]
Adding related issue 

https://github.com/tensorflow/tensorflow/issues/28942

**Comment by samhodge** [comment]
Adding related issue
https://github.com/tensorflow/tensorflow/issues/29042

**Comment by samhodge** [comment]
This issue has been reported 2 years ago

https://github.com/bazelbuild/bazel/issues/1920

**Comment by samhodge** [comment]
Adding solution to issue reported 2 years ago

https://github.com/bazelbuild/bazel/issues/5200

**Comment by samhodge** [comment]
Where are the best resources to start to learn how to do this myself in Skylark?

**Comment by gunan** [comment]
@hlopko @meteorcloudy to help with starlark documentation.

**Comment by samhodge** [comment]
I might add that Tensorflow/contrib/cmake is very broken under Linux for v1.12.0 and v1.13.1 the abseil-cpp dependency is missing from v1.12.0 and in addition v1.12.1+ has a dnn.proto file that should get converted to a dnn.pb.h but is not included in any dependency. @gunan Do you know if there is a maintainer of contrib/cmake they seem like relatively easy fixes for someone with good cmake skills. The other issue were some files in v1.12.0 had been renamed or moved or missing from lists.

**Comment by gunan** [comment]
We have dropped official support for cmake long ago. To the best of my
knowledge, noone stepped up to claim ownership of the cmake folder. That
said, I am still happy to accept fixes for the cmake files.

**Comment by samhodge** [comment]
@gunan 

Thanks for your response, my own CMake skills are limited to a number of files in a folder and linking to a library or two, rather than checking out whole projects and linking them together from a single CMake parent file.

That being said at some stage in my life I didn't know how to walk or talk, and I learned that too. My choice now is if to make the small number of changes via CMake or the other changes via bazel and Skylark. I really need static linking with CUDA support to satisfy the use case for Tensorflow as a plugin to a parent package already using Tensorflow where computations are in the minutes per frame CPU only but are seconds a frame with GPU support.

 But there is plenty of example code and tutorials at least for CMake. So I may make those contributions.

**Comment by samhodge** [comment]
Ok CMake seems like a lost cause.

Although in October 2018 I did build a static version of Tensorflow 1.8(?) on Windows with CUDA 9 but that is not suitable to current RTX hardware 

So looking into Bazel and Starlark

I found this

https://docs.bazel.build/versions/0.26.0/be/c-cpp.html

With reference to link static 

With a bit of graph introspection of libtensorflow_cc.so

I can probably expand out the transitive libraries 

But I don’t know how to make an archive from an existing statically linked archive, are there any tips on how this is done?

Sam

**Comment by hlopko** [comment]
You'll have to unarchive all archives and then re-archive all objects back into 1 giant archive. Or use ar script (https://stackoverflow.com/questions/3821916/how-to-merge-two-ar-static-libraries-into-one).

**Comment by samhodge** [comment]
@hlopko

Is there a way to do the ar unpack repack trick in Skylark?

Or would it be easier to do some graph walking and just stuff all of the .o files for recursive upstream dependencies to libtensorflow_cc.so?

I am hoping to have time to tackle this today.

**Comment by samhodge** [comment]
I found how to get the ar executable 

https://docs.bazel.build/versions/master/skylark/lib/CcToolchainInfo.html

But not sure how to call it

@hkoplo is there an example in the docs or could you paste a snippet

**Comment by samhodge** [comment]
Seems the ar actions take flag

https://docs.bazel.build/versions/master/cc-toolchain-config-reference.html#ar-actions

So you could call ar -x on each .a file initially

But where will the .o files end up? Also what if the names of the object files overlap?

**Comment by samhodge** [comment]
So the journey begins

```
Kognats-Mac-Pro:tensorflow kognat$ bazel query 'deps(//tensorflow:libtensorflow_cc.so)' --output graph > graph.in
Starting local Bazel server and connecting to it...
ERROR: /Users/kognat/dev/tensorflow-v1.13.1-static-gpu/tensorflow/tensorflow/cc/saved_model/BUILD:36:1: no such target '//tensorflow/core:saved_model_portable_proto': target 'saved_model_portable_proto' not declared in package 'tensorflow/core' defined by /Users/kognat/dev/tensorflow-v1.13.1-static-gpu/tensorflow/tensorflow/core/BUILD and referenced by '//tensorflow/cc/saved_model:reader'
ERROR: Evaluation of query "deps(//tensorflow:libtensorflow_cc.so)" failed: errors were encountered while computing transitive closure
```

Just trying

bazel build --config=opt //tensorflow:libtensorflow_cc.so

to be sure it can actually still be built, will know more in an hour or two

here is my .tf_configure.bazelrc

```
Kognats-Mac-Pro:tensorflow kognat$ cat .tf_configure.bazelrc 
build --action_env PYTHON_BIN_PATH="/Users/kognat/anaconda3/envs/tensorflow/bin/python"
build --action_env PYTHON_LIB_PATH="/Users/kognat/anaconda3/envs/tensorflow/lib/python3.6/site-packages"
build --python_path="/Users/kognat/anaconda3/envs/tensorflow/bin/python"
build:xla --define with_xla_support=true
build --action_env TF_NEED_OPENCL_SYCL="0"
build --action_env TF_NEED_ROCM="0"
build --action_env TF_NEED_CUDA="1"
build --action_env CUDA_TOOLKIT_PATH="/usr/local/cuda"
build --action_env TF_CUDA_VERSION="10.0"
build --action_env CUDNN_INSTALL_PATH="/usr/local/cuda"
build --action_env TF_CUDNN_VERSION="7"
build --action_env TF_CUDA_COMPUTE_CAPABILITIES="6.1"
build --action_env TF_CUDA_CLANG="0"
build --action_env GCC_HOST_COMPILER_PATH="/usr/bin/gcc"
build --config=cuda
test --config=cuda
build:opt --copt=-march=native
build:opt --copt=-Wno-sign-compare
build:opt --host_copt=-march=native
build:opt --define with_default_optimizations=true
build:v2 --define=tf_api_version=2
```

**Comment by samhodge** [comment]
compile fails

trying again with 
```
bazel build --config=opt --config=nonccl //tensorflow:libtensorflow_cc.so
```

**Comment by samhodge** [comment]
Also building on CentOS 7.4 from v1.14.0

With the following config

```
[kognat@vxfhost tensorflow]$ more .tf_configure.bazelrc 
build --action_env PYTHON_BIN_PATH="/usr/bin/python3"
build --action_env PYTHON_LIB_PATH="/usr/lib64/python3.6/site-packages"
build --python_path="/usr/bin/python3"
build:xla --define with_xla_support=true
build --action_env TF_CUDA_VERSION="10"
build --action_env TF_CUDNN_VERSION="7"
build --action_env TF_CUDA_PATHS="/usr/local/cuda-10.0,/usr/local/cudnn-7.4.2.24/cuda,/usr/local/cuda-10.0/extras
/CUPTI/"
build --action_env CUDA_TOOLKIT_PATH="/usr/local/cuda-10.0"
build --action_env TF_CUDA_COMPUTE_CAPABILITIES="6.1"
build --action_env LD_LIBRARY_PATH="/usr/local/cuda-10.0/lib64/:/usr/local/cuda-10.0/extras/CUPTI/lib64/:/usr/local/cudnn-7.4.2.24/cuda/lib64/"
build --action_env GCC_HOST_COMPILER_PATH="/usr/lib64/ccache/gcc"
build --config=cuda
build:opt --copt=-march=native
build:opt --copt=-Wno-sign-compare
build:opt --host_copt=-march=native
build:opt --define with_default_optimizations=true
build:v2 --define=tf_api_version=2
test --flaky_test_attempts=3
test --test_size_filters=small,medium
test --test_tag_filters=-benchmark-test,-no_oss,-oss_serial
test --build_tag_filters=-benchmark-test,-no_oss
test --test_tag_filters=-no_gpu
test --build_tag_filters=-no_gpu
test --test_env=LD_LIBRARY_PATH
build --action_env TF_CONFIGURE_IOS="0"
```

using
bazel build --config=opt --action_env=LD_LIBRARY_PATH --config=cuda //tensorflow:libtensorflow_cc.so

where CUDNN, CUDA TOOLKIT and CUPTI extra are all put on the LD_LIBRARY_PATH

with updated git 2.16
gcc/4.8.5

Cuda Toolkit 10.0
CUDNN 7.4.2.24

Once that is complete I will be able to start messing with repacking the symbols with edits to the tensorflow.bzl

but

initially the query above did not work.

**Comment by samhodge** [comment]
Adding to the list of things that dont work, doing 
```
find bazel-out -iname \*.o
```

mashing that though
```
ar -r libtfgpu.a lots.pic.o of.pic.o things.o
```

Expecting to use 
```
-Wl,whole_archive -Wl,--whole-archive $FOLDER/libtfgpu.a -Wl,--no-whole-archive  and expect 
```

everything to be OK

But it was worth a shot.

**Comment by samhodge** [comment]
I have attached the graph that resolves with the query

 bazel query --noimplicit_deps 'deps(//tensorflow:libtensorflow_cc.so)' --output graph > graph.in

after deleting a few lines from 
```
tensorflow/cc/saved_model/BUILD
```

[graph.txt](https://github.com/tensorflow/tensorflow/files/3268055/graph.txt)

It is pretty big.

**Comment by samhodge** [comment]
I gave up and used the Golidlocks method with the .o files making a .a file that was just right.

Basically you play godilock with the intermediate archive.

Go to link and it says "Symbol Missing" you add more .o files

Add too many .o files and you get "Duplicate Symbol"

Until like Goldilocks you find a list of symbols that is "Just Right"

Then you prorgram will run the same as being Dynamically Linked.

As a result I still dont know much Starlark.

On OSX and Linux a list of those object files for CUDA and v1.12.0 and my configuration is 

[Linux_object.txt](https://github.com/tensorflow/tensorflow/files/3269863/Linux_object.txt)
[OSX_objs.txt](https://github.com/tensorflow/tensorflow/files/3269864/OSX_objs.txt)

**Comment by royhershkovitz** [comment]
https://www.tensorflow.org/lite/guide/build_arm64
who need it for arm can do it this way.

**Comment by WTRipper** [comment]
@samhodge Did you find a less direct method to build a static library? If not could you explain your approach a bit more in detail? which tools did you use exactly?  
I am currently struggling with creating a static library for windows.

**Comment by royhershkovitz** [comment]
I am still struggling with it for Arm, it worked well with CMake on Linux x86, but the arch in Arm file is not fine for the cross compilers.  The next approach will be compiling everything with bazel.

**Comment by samhodge** [comment]
I recently revisited this as part of the bazel process in bazel-bin directory there are _obj folders with the .o files

Running a 

find bazel-bin -name “*.o” | grep -v “not these” > list.txt

Will get you a list of all of the .o files to make the bazel target you chose like libtensorflow_cc.so

Then you can write a short Python command to turn that list into a sub process to pack those .o files into a .a via ar with the appropriate flags.

Later when you are linking your application you might need so other symbols: absl, protobuf, nsync.

But that is the gist of my approach.

Sam

**Comment by svking** [comment]
> Any progress on this, I tried hacking up the Tegra gpu/cuda libtfcuda.a target to work on Linux and I didn't get very far:


Did you compile libtfcuda.a  successfully ？

**Comment by svking** [comment]
> I recently revisited this as part of the bazel process in bazel-bin directory there are _obj folders with the .o files
> 
> Running a
> 
> find bazel-bin -name “*.o” | grep -v “not these” > list.txt
> 
> Will get you a list of all of the .o files to make the bazel target you chose like libtensorflow_cc.so
> 
> Then you can write a short Python command to turn that list into a sub process to pack those .o files into a .a via ar with the appropriate flags.
> 
> Later when you are linking your application you might need so other symbols: absl, protobuf, nsync.
> 
> But that is the gist of my approach.
> 
> Sam

Does the .a file  depend on libcuda.so and other .so？

**Comment by kognat-docs** [comment]
yes

**Comment by samhodge** [comment]
> > I recently revisited this as part of the bazel process in bazel-bin directory there are _obj folders with the .o files
> > Running a
> > find bazel-bin -name “*.o” | grep -v “not these” > list.txt
> > Will get you a list of all of the .o files to make the bazel target you chose like libtensorflow_cc.so
> > Then you can write a short Python command to turn that list into a sub process to pack those .o files into a .a via ar with the appropriate flags.
> > Later when you are linking your application you might need so other symbols: absl, protobuf, nsync.
> > But that is the gist of my approach.
> > Sam
> 
> Does the .a file depend on libcuda.so and other .so？

yes it does, if you compile with CUDA support.

**Comment by arlofaria** [comment]
I was able to make some things work by using the object files reported in the `*.params`:
```
ar -cr libtensorflow.a $(cat bazel-bin/tensorflow/libtensorflow_*.so.*.params | grep '\.o$')
````
... and then link with `-Wl,--allow-multiple-definition -Wl,--whole-archive libtensorflow.a`

Note: I first tried creating separate static libraries corresponding to the shared libraries, `libtensorflow_cc.a` and `libtensorflow_framework.a` as noted in this issue's title, but encountered a lot of problems with that approach (e.g. duplicate registrations).  Putting it all in one archive seemed to magically resolve that.

Hope this helps someone!

**Comment by arlofaria** [comment]
Actually, it's better with `bazel build --config=monolithic`, which seems intended for static linking and enables:
`ar -cr libtensorflow_cc.a $(cat bazel-bin/tensorflow/libtensorflow_cc.so.*.params | grep '\.o$')`

**Comment by loadingvx** [comment]
`monolithic` contains only one option define which was

`build:monolithic --define framework_shared_object=false`

And it is default true, so don't worry about `--config=monolithic` option, and you don't need to recompile with this option if you don't know it. (for any one who likes me, and I'm compiling tfv2.6)

```
# For projects which use TensorFlow as part of a Bazel build process, putting
# nothing in a bazelrc will default to a monolithic build. The following line
# opts in to modular op registration support by default.
build --define framework_shared_object=true
```

**Comment by kumariko** [comment]
We see that you are using old version of tensorflow which is out of support window, We recommend that you upgrade to 2.6.0 and let us know if the issue still persists in newer versions .please check https://github.com/tensorflow/tensorflow/issues/28388#issuecomment-733962842 , https://github.com/tensorflow/tensorflow/issues/28388#issuecomment-734028952 it will help in resolving your issue. Please open a new issue in case you face any errors, we will get you the right help .Hence moving this to closed status.Thanks!

**Comment by Frinklin-Wang** [comment]
> I sorted the issues on the Linux side of things and can get everything statically linked with libtensorflow-core.a nsync.a and libprotobuf.a I was missing some of the -W,l directives which I shamelessly stole from the benchmark example that came as part of contrib/makefile
> 
> EDIT:
> 
> here is the info
> 
> ```
> -Wall -L$(TENSORFLOW)/tensorflow/contrib/makefile/gen/protobuf-host/lib -Wl,--allow-multiple-definition -Wl,--whole-archive  $(TENSORFLOW)/tensorflow/contrib/makefile/gen/lib/libtensorflow-core.a -Wl,--no-whole-archive $(TENSORFLOW)/tensorflow/contrib/makefile/downloads/nsync/builds/default.linux.c++11/nsync.a -lstdc++ -l:libprotobuf.a -lz -lm -ldl -lpthread -lrt 
> ```

Hi @samhodge ,
Now, I build the c++ static library . I want to have a test, but there are some errors when I link the libs.
**Describe the problem:**  
**#62868** 
Could you please give some help? Thank you!
